backend/src/services/course/course.service.ts

```diff
@@ -0,0 +1,76 @@
+import { CourseHost, ExceptionMessage, VendorKey } from '~/common/enums/enums';
+import {
+  CourseCreateArgumentsDto,
+  CourseGetResponseDto,
+} from '~/common/types/types';
+import { course as courseRep } from '~/data/repositories/repositories';
+import { CoursesError } from '~/exceptions/exceptions';
+import { udemy as udemyServ, vendor as vendorServ } from '~/services/services';
+
+type Constructor = {
+  courseRepository: typeof courseRep;
+  vendorService: typeof vendorServ;
+  udemyService: typeof udemyServ;
+};
+
+class Course {
+  #courseRepository: typeof courseRep;
+  #vendorService: typeof vendorServ;
+  #udemyService: typeof udemyServ;
+
+  constructor({ courseRepository, vendorService, udemyService }: Constructor) {
+    this.#courseRepository = courseRepository;
+    this.#vendorService = vendorService;
+    this.#udemyService = udemyService;
+  }
+
+  async create(
+    courseRequestDto: CourseCreateArgumentsDto,
+  ): Promise<CourseGetResponseDto> {
+    const { description, title, url, vendorKey } = courseRequestDto;
+
+    const vendor = await this.#vendorService.getByKey(vendorKey);
+
+    if (!vendor) {
+      throw new CoursesError({
+        message: ExceptionMessage.INVALID_COURSE_VENDOR,
+      });
+    }
+
+    const course = await this.#courseRepository.create({
+      description,
+      title,
+      url,
+      vendorId: vendor.id,
+    });
+
+    return course;
+  }
+
+  async createByUrl(url: string): Promise<CourseGetResponseDto | null> {
+    const urlObject = new URL(url);
+    const { host } = urlObject;
+
+    switch (host) {
+      case CourseHost.UDEMY:
+      case CourseHost.W_UDEMY: {
+        const courseData = await this.#udemyService.getByUrl(urlObject);
+        const { description, title, url } = courseData;
+
+        return this.create({
+          description,
+          title,
+          url,
+          vendorKey: VendorKey.UDEMY,
+        });
```

Comment:
we can create duplicates, @what1s1ove it's ok for now?

Reply:
Regarding this, I couldn't really think of the best way to avoid it.
The one solution could be by checking `url`, but even in that case, we can have at least 2 duplicates: 
udemy.com/:courseId, udemy.com/:courseSlug
Another one I thought about is by fetching the `id` from the API and saving it, but I don't think we should fetch the `id`

Reply:
the id is only thing that won't change and we can rely on it. Why you think we shouldn't use the id?

Reply:
By "shouldn't" I meant because it's not in DB schema, but surely, it can be fetched and used for that purpose. Should I do it?

Reply:
We'll do this as separate ticket 👍